angular/projects/spark-angular/src/lib/components/sprk-tooltip/sprk-tooltip.component.ts

```diff
@@ -0,0 +1,254 @@
+   * library to capture data.
+   */
+  @Input()
+  analyticsString: string;
+  /**
+   * Expects a space separated string
+   * of classes to be added to the
+   * tooltip element.
+   */
+  @Input()
+  additionalClasses: string;
+  /**
+   * Expects a space separated string
+   * of classes to be added to the
+   * svg icon.
+   */
+  @Input()
+  iconAdditionalClasses: string;
+  /**
+   * The value supplied will be assigned
+   * to the `data-id` attribute on the
+   * container element. This is intended to be
+   * used as a selector for automated
+   * tools. This value should be unique
+   * per page.
+   */
+  @Input()
+  idString: string;
+  /**
+   * Optional: the unique ID to use for the tooltip element.
+   * If an ID is not provided, a unique ID will be created
+   * automatically.
+   */
+  @Input()
+  id = uniqueId(`sprk_tooltip_`);
+
+  /**
+   * Emitted when the tooltip is toggled open.
+   */
+  @Output()
+  openedEvent = new EventEmitter<any>();
+
+  /**
+   * Emitted when the tooltip is toggled closed.
+   */
+  @Output()
+  closedEvent = new EventEmitter<any>();
+
+  /**
+   * @ignore
+   */
+  @ViewChild('triggerElement') triggerElement;
+
+  /**
+   * @ignore
+   */
+  @ViewChild('tooltipElement') tooltipElement;
+
+  /**
+   * @ignore
+   */
+  @ViewChild('containerElement') containerElement;
+
+  /**
+   * @ignore
+   */
+  @HostListener('document:keydown', ['$event'])
+  onKeydown($event) {
+    if (this.isToggled) {
+      if ($event.key === 'Escape' || $event.key === 'Esc' || $event.keyCode === 27) {
+        this.isToggled = false;
+        this.closedEvent.emit();
+      }
+    }
+  }
+
+  /**
+   * @ignore
+   */
+  @HostListener('document:click', ['$event'])
+  onDocumentClick(event): void {
+    if (this.containerElement && !this.containerElement.nativeElement.contains(event.target)) {
+      if (this.isToggled) {
+        this.isToggled = false;
+        this.closedEvent.emit();
+      }
+    }
+  }
+
+  /**
+   * @ignore
+   */
+  @HostListener('focusin')
+  onFocusIn() {
+    this.setPositioningClass();
+  }
+
+  /**
+   * @ignore
+   */
+  @HostListener('mouseover')
+  onMouseOver() {
+    this.setPositioningClass();
+  }
+
+  /**
+   * @ignore
+   */
+  @HostListener('click', ['$event'])
+  onClick(event) {
+    this.toggle()
+  }
+
+  /**
+   * @ignore
+   */
+  public positioningClass = 'sprk-c-Tooltip--bottom-right';
+
+  /**
+   * @ignore
+   */
+  setPositioningClass(): void {
+    var trigger = this.triggerElement.nativeElement;
+
+    const elemX = trigger.getBoundingClientRect().left;
+    const elemY = trigger.getBoundingClientRect().top;
+
+    const viewportWidth = window.innerWidth;
+    const viewportHeight = window.innerHeight;
+
+    // 328 is the default max-width
+    var maxWidth = 328;
+    var calculatedWidth = maxWidth;
+
+    if (elemX > viewportWidth / 2) {
+      // the left edge of the button + the width of the button + the border
+      calculatedWidth = elemX + 16 + 16;
```

Comment:
Lets make any numerical value a variable. Doing that, removes the need for comments explaining what it is because the variable name will explain that.